### PR TITLE
Support for long-contexts with LoRA

### PR DESCRIPTION
**Pull request by SanjuCSudhakaran**
This PR does the following

1. Enable long-context support for LoRA
2. Handle RoPE changes requires for LoRA. The corresponding vllm-hpu-extension changes can be found https://github.com/HabanaAI/vllm-hpu-extension

### CODE REVIEW

File: vllm/lora/punica.py [hunk 1]
```diff
@@ -102,11 +102,8 @@ def convert_mapping(
     index_mapping_indices: List[int] = list(mapping.index_mapping).copy()
     embedding_indices = index_mapping_indices.copy()
     lora_indices = index_mapping_indices.copy()
-    long_lora_offsets: Optional[torch.Tensor] = None
-    if long_lora_context:
-        long_lora_offsets = torch.zeros(len(index_mapping_indices),
-                                        device=get_device(),
-                                        dtype=torch.long)
+    long_lora_offsets_list = []
```
Comment (vivekgoe) [inline vllm/lora/punica.py:105]: follow notation, add type List[int]

File: vllm/lora/punica.py [hunk 2]
```diff
@@ -119,11 +116,13 @@ def convert_mapping(
         embedding_indices[i] = lora_idx if index_mapping_indices[i] > 0 else 0
         lora_indices[i] = lora_idx
         if long_lora_context:
-            assert long_lora_offsets is not None
             lora_offset: int = long_lora_context.offsets_by_lora_id.get(
                 index_mapping_indices[i], 0)
-            long_lora_offsets[i] = lora_offset
+            long_lora_offsets_list.append(lora_offset)
 
+    long_lora_offsets = torch.tensor(long_lora_offsets_list,
```
Comment (vivekgoe) [inline vllm/lora/punica.py:123]: W.r.t. functionality this change is ok for other backends (e.g. CUDA) also. @kzawora-intel is it ok to keep it? or do we need to make it HPU specific to be more upstream(ing) friendly?

File: vllm/lora/layers.py [hunk 3]
```diff
@@ -1296,12 +1309,8 @@ def forward(
         query: torch.Tensor,
         key: torch.Tensor,
     ) -> Tuple[torch.Tensor, torch.Tensor]:
-        return self.base_layer(
-            positions,
-            query,
-            key,
-            offsets=self.punica_wrapper.long_lora_indices,
-        )
+        offsets = self.punica_wrapper.long_lora_indices.reshape_as(positions)
```
Comment (vivekgoe) [inline vllm/lora/layers.py:1312]: Is this change ok for other backends (e.g. CUDA)?

File: vllm/lora/layers.py [hunk 4]
```diff
@@ -1316,8 +1325,12 @@ def can_replace_layer(
         model_config: Optional[PretrainedConfig],
     ) -> bool:
         """Returns True if the layer can be replaced by this LoRA layer."""
-        return (type(source_layer) is LinearScalingRotaryEmbedding
-                or type(source_layer) is RotaryEmbedding)
+        if current_platform.is_hpu():
+            return (type(source_layer) is HpuLinearScalingRotaryEmbedding
```
Comment (vivekgoe) [inline vllm/lora/layers.py:1329]: Is there any functional impact of using default code?

**Review by vivekgoe — CHANGES_REQUESTED** [review]
@SanjuCSudhakaran added review comments, please check.

**Comment by SanjuCSudhakaran** [comment]
These changes are no longer required since the latest vllm-fork already refactored RoPE to handle this.

PR https://github.com/HabanaAI/vllm-fork/pull/404

**Pull request closed without merging**